### DISCUSSION TITLE
白嫖 ChatGPT 的方案

### DISCUSSION THREAD

**Discussion by adoyle-h** [just-talks/tech-talks#23] [4.Ideas]
1. 到 [OpenAPI 官网](https://platform.openai.com/)注册新账号，它会送你 ~~$18~~ $5，限时 3 个月。只要不是很多人用，$5 也足够了。
2. 到 [Setting 页面生成  secret API key](https://platform.openai.com/account/api-keys)。
3. 使用 https://github.com/Chanzhaoyu/chatgpt-web 在 [Railway 直接部署服务](https://github.com/Chanzhaoyu/chatgpt-web#%E4%BD%BF%E7%94%A8-railway-%E9%83%A8%E7%BD%B2)
    - 建议把网页 index.html 文件的 title 改了，以及加个 robots.txt 禁止被 SEO 收录。可以参考[我的修改](https://github.com/adoyle-h/chatgpt-web/commit/c709f8de8267378437ad4183bfb95159a5be73a7)。因为我怀疑 GFW 会根据搜索引擎找到网站从而域名被墙。参考[这个评论](https://github.com/just-talks/tech-talks/discussions/23#discussioncomment-5499689)。
    - **注意，如果你要部署到自己的机器，机器必须得在国外。在国内的话，你的 OpenAI 账号极大概率会被封。**
4. 部署成功后在 Railway 的应用设置里打开域名 (domain)，Railway 会分配一个子域名给你。
    - 你也可以添加自定义域名。如果你用 Cloudflare，记得域名的 SSL/TLS 设置必须是 `Full`（默认 `Flexible`），否则会导致前端重定向过多 `ERR_TOO_MANY_REDIRECTS`。
5. 访问成功。

Railway 部署体验非常棒。流畅丝滑。十五分钟内搞定所有。

![CleanShot 2023-03-03 at 15 48 23@2x](https://user-images.githubusercontent.com/1998490/222672023-9f8754c8-50e9-4955-8b9f-e1bcdb91bc51.png)

<img width="1343" alt="CleanShot 2023-03-03 at 16 40 11@2x" src="https://user-images.githubusercontent.com/1998490/222672859-54e016d6-5e64-436c-8d80-a7c5fd2671b7.png">

Railway 可以不填写信用卡就能用，但限制是限时 500 小时（21 天）。基础套餐每月免费送 $5。就自己用的话 $5 肯定够了，主要是服务时间。只要服务部署完毕就开始计时，无论是否有流量。可以周末或晚上关闭服务来节省时间，就能完全白嫖了。

但如果填写了信用卡，可以升级到 Developer Plan，相比基础套餐的区别就是取消了 500 小时限制。依然每月免费送 $5。超出金额的部分会通过信用卡扣费。具体看[它的 Plan](https://docs.railway.app/reference/plans)。

这是我一个月只开 chatgpt-web 的用量。可以参考。

<img width="910" alt="CleanShot 2023-03-31 at 10 15 37@2x" src="https://user-images.githubusercontent.com/1998490/229005822-1db9c6bd-9f4b-41b0-81eb-26af310e1c06.png">

Railway 不支持停止服务，只支持移除服务。只要按 remove 按钮就能停止，然后按 rollback 按钮重启就行了，其他设置都保留的。

---
2023/03/15 更新

ChatGPT 4 发布啦，可以调用 OpenAPI，不过需要先[提交申请](https://openai.com/waitlist/gpt-4-api)。

https://github.com/Chanzhaoyu/chatgpt-web 支持切换模型，加一个 `OPENAI_API_MODEL=gpt-4` 就好啦。

<img width="843" alt="CleanShot 2023-03-15 at 13 30 09@2x" src="https://user-images.githubusercontent.com/1998490/225215957-28517dc7-11b1-4ed5-a071-d54809d2dd9f.png">

---
2023/03/31 更新

更新到最新版本的 chatgpt-web。它支持设置 `AUTH_SECRET_KEY` 变量添加你的密码访问权限。就不怕被别人刷爆卡啦。

---
2023/04/03 更新

过期后继续使用则需要绑定信用卡开通 paid plan。你可以去开一个美国虚拟信用卡。有几个注意点：

1.  虚拟卡的信息要和开通 OpenAI paid plan 的信息要保持一致，包括姓名、手机号、地址。全是假的没关系。用[美国地址生成器](https://www.meiguodizhi.com/)生成。
2. 地址建议选免税州 ：蒙大拿州（Montana）、俄勒冈州（Oregon）、阿拉斯加州（Alaska）、特拉华州（Delaware）、新罕布什尔州（New Hampshire）
3. 根据网上其他人的分享，以下信用卡是支持的，6 位数字是卡段（卡号前缀）

- DePay：531993
- NobePay：556766、556735、531847
- Payoneer：529366
- QBT 虚拟卡：489683、540524、556150
-  485953、428816（我用这个成功的）

4. 以下信用卡不支持

- 国内招行 VISA/MASTERCARD、全币卡
- 全球付香港虚拟卡
- 万里汇香港虚拟卡

5. 虚拟卡的有效期都很短，半年到一年，过期了又要重新开卡。开卡费几十块。
6. 可以参考 [keroro 的评论](https://github.com/just-talks/tech-talks/discussions/23#discussioncomment-5294230)。
7. 在 OpenAI 网站上操作时，IP 最好是美国的。别是高风险的 ip，你可以用 https://scamalytics.com/ip 来测试一下，Fraud Score 是 100 的就不要用这个 ip，我的 Fraud Score 是 69，可以成功绑定。

**Comment by adoyle-h** [comment]
Q: 这 3 个月后不就得付费了吗？
A: 要这么看，ChatGPT 3 OpenAPI 的费用也就 $0.002 每一千单词， 100 万个单词才 18 RMB（一杯奶茶或咖啡的钱）。自己部署的服务比用他人二手封装的服务，相对更安全一点。

费用 = 提示词价格*提示词数量（用户输入的） + 补全词价格*补全词数量（ChatGPT 返回的）

ChatGPT 4 发布后，价格是：
ChatGPT 4 （8K 上下文窗口，大约可存储 13 页文本的上下文）：$0.03 每一千提示词 + $0.06 每一千补全词。
ChatGPT 4 （32K 上下文窗口，大约可存储 52 页文本的上下文）：$0.06 每一千提示词 + $0.12 每一千补全词。

Q: 用 https://chat.openai.com/chat 不是更香吗？
A: 官方免费版响应速度太慢了，体验过快速响应的 chatgpt，你就不想再用旧的方案了。

**Comment by kalxd** [comment]
卡在第一步。

**Reply by adoyle-h** [reply]
什么问题？

**Reply by kalxd** [reply]
翻墙和虚拟手机号。

**Reply by adoyle-h** [reply]
翻墙方法自己搜索。
虚拟手机号用这个 https://sms-activate.org/

**Comment by keroro520** [comment]
使用 https://github.com/Chanzhaoyu/chatgpt-web 在 [Railway 直接部署服务](https://github.com/Chanzhaoyu/chatgpt-web#%E4%BD%BF%E7%94%A8-railway-%E9%83%A8%E7%BD%B2)，真是太丝滑了。我这种没听过 Railway 的小白都很快就部署好了。感谢

**Reply by adoyle-h** [reply]
Railway 的缺点是每月免费只送 $5 和 500 小时（21 天）。就自己用的话 $5 肯定够了，主要是服务时间。只要服务部署完毕就开始计时，无论是否有流量。可以周末或晚上关闭服务来节省时间，就能完全白嫖了。
Railway 不支持停止服务，只支持移除服务。只要按 remove 按钮就能停止，然后按 rollback 按钮重启就行了，其他设置都保留的。

**Reply by adoyle-h** [reply]
发现原来填信用卡开通 developer plan，就能移除 500 小时限制。而且照样每月送 $5，超出部分从信用卡扣费。我这个月只用了 $1.09。挺香的。更新到原帖了，你可以看看。

**Reply by keroro520** [reply]
已开通 developer plan 。感谢分享！  @adoyle-h 

But a bad news: 我在 Railway 上搭建的服务，原来不需要翻墙的，现在必须要翻墙了。我的初衷是分享给国内没用过梯子的朋友们啊 😢

**Reply by adoyle-h** [reply]
我的域名前天也被封了。不知道是什么原因被动触发了 GFW，还是 GFW 主动扫描的，我觉得后者可能性更高。发现也有其他人最近被墙的，看这个 [issue](https://github.com/Chanzhaoyu/chatgpt-web/issues/1026)。

不过我发现它没有封 railway 的域名，你可以访问 railway 提供的那很长的域名。
或者如果你不担心二级域名被墙的话，再开一个新的三级域名。我发现 GFW 只墙了当前的三级域名，没墙二级域名。
或者新买个便宜的国外域名，首年优惠什么的，被墙也不心疼。

另外建议把网页 index.html 文件的 title 改了，以及加个 robots.txt 禁止被 SEO 收录。可以参考[我的修改](https://github.com/adoyle-h/chatgpt-web/commit/c709f8de8267378437ad4183bfb95159a5be73a7)。我怀疑 GFW 是根据搜索引擎找到了我的网站。
因为我的国外域名用了很多年从来没被墙过，而且 GPT web 用了一个月也没事，就前天突然被墙。而其他三级域名都没事。

最令我头疼的不是墙的问题，而是我的 openai 的免费额度到期了。它不支持国内信用卡，必须要用国外的信用卡。日本和香港的都不行，美国的可以。网上有很多开通美国虚拟信用卡的方案，我都看了一遍，感觉成本太高了，开卡费用要充值 $30。我用 ChatGPT 可能要用一年才用完。但指不定这些平台就卷钱跑路了。

**Reply by keroro520** [reply]
我也留言一下最近折腾 ~~充值~~ ChatGPT-Plus 被 OpenAI 的风控搞得崩溃。

因为最近 OpenAI 的风控升级了，我用Depay的卡也弄不了，提示 your card has been declined。

看网上的经验，主要还是ip的问题，商业梯子的ip容易被风控，反正我买的梯子估计都被黑名单了。
另外我在AWS开了台Windows，通过远程桌面的方式去操作 pay.openai.com 我是如此提示。我都怀疑是针对 depay card number 的风控了。

还没折腾成功。

我现在在 Azure 上申请 OpenAI API service https://twitter.com/zhangjintao9020/status/1641688163587588099?s=19 等结果

**Reply by keroro520** [reply]
已成功在 platform.openai.com 上绑定了 Depay 卡，至少 API keys 还可以继续用了。

但是 ChatGPT-Plus 依旧失败在 your card has been declined

**Reply by keroro520** [reply]
终于绑定上了。记录了一下经验。

## your card has been declined

貌似最近 pay.openai.com 针对梯子的 IP 做了严格的风控，这些梯子的 IP 上了风控名单后，你再基于这些 IP 去绑卡，就会报错 "your card has been declined"。

### 通过远程桌面绕过 pay.openai.com 的风控 your card has been declined

首先，你得有个 ChatGPT 账户、美国卡或者虚拟卡（我用 Depay），这些东西网上已经有很多资料了。

1. 在 AWS EC2 上启动一个 Windows 实例，并通过 RDP 连接
2. 使用远程 Windows 实例中的浏览器打开 chat.openai.com 会被禁止访问（Error 1020 You do not have access to chat.openai.com.The site owner may have set restrictions that prevent you from accessing the site）。所以，先在自己的本地打开 chat.openai.com，点击 upgrade to chatgpt-Plus，它会跳出绑定银行卡的页面（页面的域名是 https://pay.openai.com/ ）
3. 把绑定信用卡的页面的链接贴到远程 Windows 实例的浏览器中打开。

---

我碰见的一个乌龙情况：绑定卡时一直报 your card has been declined，就在我即将放弃之际，我偶然瞥见 Depay billing，有两单失败的扣款账单，竟然是 openai 扣款失败的账单，原因是余额不足。。。原来扣款失败也是报 your card has been declined 啊。所以，请确保你的卡上至少有 20u。

**Comment by adoyle-h** [comment]
ChatGPT 4 发布啦，可以调用 OpenAPI，不过需要先[提交申请](https://openai.com/waitlist/gpt-4-api)。

https://github.com/Chanzhaoyu/chatgpt-web 支持切换模型，加一个 `OPENAI_API_MODEL=gpt-4` 就好啦。

<img width="843" alt="CleanShot 2023-03-15 at 13 30 09@2x" src="https://user-images.githubusercontent.com/1998490/225215957-28517dc7-11b1-4ed5-a071-d54809d2dd9f.png">

**Comment by jiaozillj** [comment]
railway 不能达成1.要求github账号注册半年以上允许验证 2.银行卡的五位邮编
所以有其他方式替代嘛

**Reply by adoyle-h** [reply]
居然还有 github 账号的限制。你可以去淘宝或闲鱼买现成的 github 账号。
我用的时候它没有问我要银行卡或信用卡，这应该不是硬性要求。

其他平台你可以试试 https://vercel.com/ 使用体验也是相当舒爽。
但没有现成的 vercel.json 配置，你自己研究一下如何部署。

**Comment by keroro520** [comment]
我之前使用的是 Railway 提供的子域名，然而，最近发现从国内无法连通该子域名了，似乎是 Railway 屏蔽了国内的 IP？

解决办法是设置自定义域名。我在 Cloudflare 注册了个域名，**把 SSL 设置为 strict 模式**；然后按提示加上 DNS 记录。具体细节 Railway 都有指引和[文档](https://docs.railway.app/deploy/exposing-your-app#lets-encrypt-ssl-certificates) 。

设置了自定义域名后，从国内也可以不翻墙访问服务了。

**Reply by nino89** [reply]
在cloudflare上添加dns记录之后，需要等一段时间吗？

**Reply by adoyle-h** [reply]
> 在cloudflare上添加dns记录之后，需要等一段时间吗？

等几分钟就行。

**Comment by adoyle-h** [comment]
GPT-4 API 现在付费用户也可以用了。 https://www.36kr.com/p/2333702889326089
用 chatgpt-web 的加个环境变量即可 `OPENAI_API_MODEL=gpt-4`。
GPT-4 可用的模型： https://platform.openai.com/docs/models/gpt-4
